internal/mempool/mempool.go

```diff
@@ -2159,4 +2194,456 @@ func New(cfg *Config) *TxPool {
 		staged:          make(map[chainhash.Hash]*dcrutil.Tx),
 		stagedOutpoints: make(map[wire.OutPoint]*dcrutil.Tx),
 	}
+
+	mp.miningView = &txMiningView{
+		rejected:      make(map[chainhash.Hash]struct{}),
+		txGraph:       mp.newTxDescGraph(),
+		txDescs:       []*mining.TxDesc{},
+		ancestorStats: make(map[chainhash.Hash]*mining.TxAncestorStats),
+	}
+
+	return mp
+}
+
+// txDescGraph relates a set of transactions to their respective descendents and
+// ancestors. It only stores transactions that have at least one edge relating
+// to another.
+type txDescGraph struct {
+	forEachRedeemer func(tx *dcrutil.Tx, f func(redeemerTx *mining.TxDesc))
+	childrenOf      map[chainhash.Hash]map[chainhash.Hash]*mining.TxDesc
+	parentsOf       map[chainhash.Hash]map[chainhash.Hash]*mining.TxDesc
+}
+
+// txDescFind is used to inject a transaction repository into the mining view.
+// This might either come from the mempool's outpoint map or from the mining
+// view itself.
+type txDescFind func(txHash *chainhash.Hash) *mining.TxDesc
+
+// txMiningView represents a snapshot of all transactions in mempool
+// as well as the hierarchy of transactions, as they may spend from one another
+// in the mempool.
+type txMiningView struct {
+	rejected      map[chainhash.Hash]struct{}
+	txGraph       *txDescGraph
+	txDescs       []*mining.TxDesc
+	ancestorStats map[chainhash.Hash]*mining.TxAncestorStats
+}
+
+// newTxDescGraph returns a new instance of txDescGraph. The graph is provided
+// with a wrapper over the mempool's outpoint map to determine which
+// transactions in the pool spend from ones newly added to the graph.
+func (mp *TxPool) newTxDescGraph() *txDescGraph {
+	// for a given transaction, scan the mempool to find which transactions
+	// spend it.
+	forEachRedeemer := func(tx *dcrutil.Tx, f func(redeemerTx *mining.TxDesc)) {
+		prevOut := wire.OutPoint{Hash: *tx.Hash(), Tree: tx.Tree()}
+		txOutLen := uint32(len(tx.MsgTx().TxOut))
+		for i := uint32(0); i < txOutLen; i++ {
+			prevOut.Index = i
+			if txRedeemer, exists := mp.outpoints[prevOut]; exists {
+				f(&mp.pool[txRedeemer.MsgTx().TxHash()].TxDesc)
+			}
+		}
+	}
+
+	return &txDescGraph{
+		forEachRedeemer: forEachRedeemer,
+		childrenOf:      make(map[chainhash.Hash]map[chainhash.Hash]*mining.TxDesc),
+		parentsOf:       make(map[chainhash.Hash]map[chainhash.Hash]*mining.TxDesc),
+	}
+}
+
+// MiningView returns a slice of mining descriptors for all the transactions
+// in the pool in addition to a snapshot of the current pool's transaction
+// relationships.
+//
+// This is part of the mining.TxSource interface implementation and is safe for
+// concurrent access as required by the interface contract.
+func (mp *TxPool) MiningView() mining.TxMiningView {
+	mp.mtx.RLock()
+	defer mp.mtx.RUnlock()
```

Comment:
Same here regarding defer. There is no reason to defer here as there are not multiple exit paths and it's a trivial function.

Reply:
fixed